### ISSUE TITLE
Implement incoming-set caching.

### ISSUE THREAD

**Issue by linas** [opencog/atomspace#1373]
Implement incoming-set caching. When an atomspace is too large to fit into RAM, we must leave some of it on disk.  However, it can  become important to make sure that the incoming set of an atom has been loaded into RAM i.e. into the atomspace. Currently, repeatedly loading the incoming set is very slow and wasteful, and so its important to know whether an incoming set has already been loaded or not.  Technically, this caching can be done in the atomspace, I guess...

**Comment by akolonin** [comment]
Linas, if there is an estimation to which extent conventional OS swapping is not helpful and to what extend it could be done better than that? And yes, if doing this at all, it should be done in Atomspace so Atomspace would perform like any DB, like Allegro Cache. Note, that persistent DBs with such capacity all are 10-100 times slower than pure in-memory DB because of overhead on guarding access to items that could be cached-in or cached-out.

**Comment by linas** [comment]
Yes. What can I say. I worked as an operating system kernel programmer for 5-10 years. I ported the Linux kernel to the IBM mainframe. I spent several years doing performance measurement and optimization on a web+database system. I'm very aware of the issues.
* http://linuxvm.org/community/lvepstas.html
* http://www.linuxjournal.com/article/3988
* https://en.wikipedia.org/wiki/Linux_on_z_Systems

In 2018-2022 if you want to get good performance, use SSD drives attached via PCIE-3. Its expensive.  You can also use SATA-attached SSD disks, but these are maybe 4x slower or more.  use spinning disks for backup only, not for live data.

**Comment by linas** [comment]
Anyway, the atomspace is already painfully slow,  It would be nice to totally redesign it from scratch for scalability and etc. but this is a massively huge undertaking. Its also an undertaking that will almost surely end in failure unless you have top-class system architects work on it. We have to live with what it is, for now.

Anyway, the atomspace is thread-safe; it has some lock; the atoms are immutable, and the atoms and values are references with smart pointers, which means that you can use  your favorite lock-free programming technique. https://en.wikipedia.org/wiki/Software_transactional_memory  Of course, that is much easier said than done. 

The biggest performance problem is that data access in the atomspace is insanely random, so you are almost guaranteed to have cache misses, even third-level cache misses almost all the time. Which means that all algorithms run at the speed of RAM instead of th speed (and bandwidth) of cache.  Note that modern Intel and AMD CPU's have a bandwidth to RAM that is 10x less or 20x less than bandwidth and latency to cache. The current atomspace is a poor fit for modern CPU's.

**Comment by linas** [comment]
Current idea is that this can be accomplished with a new kind of ProxyNode e.g. a CachingProxyNode or variants thereof.  The ProxyNodes allow general transformations and pipelines to be applied to AtomSpace data as it moves around.
